sympy/core/add.py

```diff
@@ -352,6 +352,43 @@ def as_coeff_add(self, *deps):
             return coeff, notrat + self.args[1:]
         return S.Zero, self.args
 
+    @cacheit
+    def as_coeff_add_deps(self, *deps):
+        """
+        Rewrite as two sums: one without terms in deps and one with.
+
+        If non-commutative, the terms without the deps only include those to
+        the left of the first non-commutative term.
+        """
+        if deps:
```

Comment:
Would the following work?
```python
c, l2 = self.args_cnc
l1, l2c = sift(c, lambda x: x.has(*deps), binary=True)
return self._from_args(*l1, commutative=True), self._from_args(*(l2c +l2), commutative=False)
```

I am not sure about caching this. Why do you think it is necessary? The other `as_coeff_Foo` routines don't do this.

Instead of creating a new method, can this 3-line version be added as a capability into `as_coeff_Add/Mul` with keyword `deps`?

Reply:
I will have to look into args_cnc, but if it does what I think (in the sense that the code is working), it will, I guess, be working. Just need to make sure that there is a zero and so on.

Caching: `as_coeff_foo` use caching. No idea if it is necessary. Just copied from there.

If the three line version works, it can be used (and even if it doesn't work). Didn't strike me to add a keyword option to `as_coeff_Foo`, but makes sense.

(A two-line version of your code can be used for `as_coeff_foo` as well, I believe.)

Reply:
I used this, but skipped setting the commutative property as I do not think it always should be like that. (It would be to use commutativity for the stuff in l2c (which really should have been the first argument, l1).

Reply:
Much more compact and nicely looking code now, thanks!